flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/security/token/HiveServer2DelegationTokenProvider.java

```diff
@@ -0,0 +1,232 @@
+         * externalized (namely if a provider/receiver throws an exception then workload must be
+         * stopped).
+         */
+        if (hiveConf == null) {
+            LOG.debug(
+                    "HiveServer2 is not available (not packaged with this application), hence no "
+                            + "hiveServer2 tokens will be acquired.");
+            return false;
+        }
+        try {
+            if (!HadoopUtils.isKerberosSecurityEnabled(UserGroupInformation.getCurrentUser())) {
+                LOG.debug(
+                        "Hadoop Kerberos is not enabled,hence no hiveServer2 tokens will be acquired.");
+                return false;
+            }
+        } catch (IOException e) {
+            LOG.debug(
+                    "Hadoop Kerberos is not enabled,hence no hiveServer2 tokens will be acquired.",
+                    e);
+            return false;
+        }
+
+        if (hiveConf.getTrimmed("hive.metastore.uris", "").isEmpty()) {
+            LOG.debug(
+                    "The hive.metastore.uris item is empty,hence no hiveServer2 tokens will be acquired.");
+            return false;
+        }
+
+        if (!kerberosLoginProvider.isLoginPossible(false)) {
+            LOG.debug("Login is NOT possible,hence no hiveServer2 tokens will be acquired.");
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public ObtainedDelegationTokens obtainDelegationTokens() throws Exception {
+        UserGroupInformation freshUGI = kerberosLoginProvider.doLoginAndReturnUGI();
+        return freshUGI.doAs(
+                (PrivilegedExceptionAction<ObtainedDelegationTokens>)
+                        () -> {
+                            Preconditions.checkNotNull(hiveConf);
+                            Hive hive = Hive.get((HiveConf) hiveConf);
+                            Clock clock = Clock.systemDefaultZone();
+                            try {
+                                LOG.info("Obtaining Kerberos security token for HiveServer2");
+
+                                String principal =
+                                        hiveConf.getTrimmed(
+                                                "hive.metastore.kerberos.principal", "");
+
+                                String tokenStr =
+                                        hive.getDelegationToken(
+                                                UserGroupInformation.getCurrentUser().getUserName(),
+                                                principal);
+                                Token<DelegationTokenIdentifier> hive2Token = new Token<>();
+                                hive2Token.decodeFromUrlString(tokenStr);
+
+                                Credentials credentials = new Credentials();
+                                credentials.addToken(hive2Token.getKind(), hive2Token);
+
+                                DelegationTokenIdentifier tokenIdentifier =
```

Comment:
The information contained in hive2Token itself is as follows. 

<img width="853" alt="dbb57aa951a61b6059f9e2aaae95ab7" src="https://github.com/apache/flink/assets/14961757/cc1636c0-235b-47cb-a17e-83993112eca4">

In HadoopFSDelegationTokenProvider, you can directly call the t.decodeIdentifier() method to get the DelegationTokenIdentifier, because ServiceLoader can load Hadoop’s DelegationTokenIdentifier class, but Hive’s DelegationTokenIdentifier class cannot be loaded by ServiceLoader, so calling decodeIdentifier() will return null. Therefore, you need to use the getTokenIdentifier method to construct Hive's DelegationTokenIdentifier

![image](https://github.com/apache/flink/assets/14961757/e4ef35bf-29f4-4bf8-93e4-5fffefd9cc07)

Reply:
Would it mean that if we create a file `META-INF/services/org.apache.hadoop.security.token.TokenIdentifier` with the content of:
```
org.apache.hadoop.hive.metastore.security.DelegationTokenIdentifier
```
would solve the issue, and the token identifier would be parsed correctly?

Reply:
I will modify the code as you suggest